csrc/multidevice/utils.cpp

```diff
@@ -138,39 +129,99 @@ int64_t numDeviceDims(const TensorView* tv) {
   if (!producer->hasDeviceMesh() && !consumer->hasDeviceMesh()) {
     return false;
   }
+
   // If device mesh are different, the Expr is resharding
-  if (!(producer->getDeviceMesh() == consumer->getDeviceMesh())) {
+  if (producer->getDeviceMesh() != consumer->getDeviceMesh()) {
     return true;
   }
-  // Create a map between producer's and consumer's IterDomains. We iterate
-  // over producer's iterdomain and compare sharding type with consumer's
-  // iterdomain
+
   const std::unordered_map<IterDomain*, IterDomain*>& p2c =
       PairwiseLogicalDomainMap(producer, consumer).mapProducerToConsumer();
-  for (auto p_id : producer->getLogicalDomain()) {
+  std::unordered_set<Val*> mapped_c_root_ids;
+  for (IterDomain* p_id : producer->getLogicalDomain()) {
     const auto i = p2c.find(p_id);
     if (i == p2c.end()) {
-      // This happens e.g. when `p_id` is squeezed or is a product of a
-      // reduction. Even if `p_id` is parallelized on DID, the dimension is
-      // size-1 and doesn't trigger resharding.
+      // This happens e.g. when `p_id` is squeezed. Even if `p_id` is
+      // parallelized on DID, the squeezed dimension is size-1 and doesn't
+      // trigger resharding.
+      continue;
+    }
+    mapped_c_root_ids.insert(i->second);
+  }
+
+  std::unordered_map<ParallelType, IterDomain*> p_parallel_type_to_id;
+  for (IterDomain* p_loop_id : producer->getLoopDomain()) {
+    const ParallelType parallel_type = p_loop_id->getParallelType();
+    if (!isParallelTypeDeviceDim(parallel_type)) {
+      continue;
+    }
+
+    const auto dependencies = IterVisitor::getInputsTo(
+        {p_loop_id},
+        {producer->getLogicalDomain().begin(),
+         producer->getLogicalDomain().end()});
+    if (std::any_of(
+            dependencies.begin(), dependencies.end(), [&](Val* dependency) {
+              return p2c.count(dependency->as<IterDomain>()) > 0;
+            })) {
+      NVF_ERROR(
+          p_parallel_type_to_id.try_emplace(parallel_type, p_loop_id).second,
+          "Found multiple loop IterDomains with the same parallel type (",
+          parallel_type,
+          "): ",
+          producer);
+    }
+  }
+
+  std::unordered_map<ParallelType, IterDomain*> c_parallel_type_to_id;
+  for (IterDomain* c_loop_id : consumer->getLoopDomain()) {
+    const ParallelType parallel_type = c_loop_id->getParallelType();
+    if (!isParallelTypeDeviceDim(parallel_type)) {
       continue;
     }
 
-    auto c_id = i->second;
-    if (p_id->getParallelType() != c_id->getParallelType() &&
-        (p_id->isDeviceDim() || c_id->isDeviceDim())) {
-      // Mismatch found
+    auto dependencies = IterVisitor::getInputsTo(
+        {c_loop_id},
+        {consumer->getMaybeRootDomain().begin(),
+         consumer->getMaybeRootDomain().end()});
+    if (std::any_of(
+            dependencies.begin(), dependencies.end(), [&](Val* dependency) {
+              return mapped_c_root_ids.count(dependency->as<IterDomain>()) > 0;
```

Comment:
Similarly, ignore the consumer's unmapped root IterDomains. These may come from DID-spliting a broadcast dimension.